go/ai/evaluator.go

```diff
@@ -0,0 +1,240 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/firebase/genkit/go/core"
+	"github.com/firebase/genkit/go/core/logger"
+	"github.com/firebase/genkit/go/core/tracing"
+	"github.com/firebase/genkit/go/internal/atype"
+	"github.com/firebase/genkit/go/internal/registry"
+	"github.com/google/uuid"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// Evaluator represents a evaluator action.
+type Evaluator interface {
+	// Name returns the name of the evaluator.
+	Name() string
+	// Evaluates a dataset.
+	Evaluate(ctx context.Context, req *EvaluatorRequest) (*EvaluatorResponse, error)
+}
+
+type (
+	evaluatorActionDef core.ActionDef[*EvaluatorRequest, *EvaluatorResponse, struct{}]
+
+	evaluatorAction = core.ActionDef[*EvaluatorRequest, *EvaluatorResponse, struct{}]
+)
+
+// Example is a single example that requires evaluation
+type Example struct {
+	TestCaseId string   `json:"testCaseId,omitempty"`
+	Input      any      `json:"input"`
+	Output     any      `json:"output,omitempty"`
+	Context    []any    `json:"context,omitempty"`
+	Reference  any      `json:"reference,omitempty"`
+	TraceIds   []string `json:"traceIds,omitempty"`
+}
+
+// Dataset is a collection of [Example]
+type Dataset = []Example
+
+// EvaluatorRequest is the data we pass to evaluate a dataset.
+// The Options field is specific to the actual evaluator implementation.
+type EvaluatorRequest struct {
+	Dataset      *Dataset `json:"dataset"`
+	EvaluationId string   `json:"evalRunId"`
+	Options      any      `json:"options,omitempty"`
+}
+
+// ScoreStatus is an enum used to indicate if a Score has passed or failed. This
+// drives additional features in tooling / the Dev UI.
+type ScoreStatus int
+
+const (
+	ScoreStatusUnknown ScoreStatus = iota
+	ScoreStatusFail
+	ScoreStatusPass
+)
+
+var statusName = map[ScoreStatus]string{
+	ScoreStatusUnknown: "unknown",
+	ScoreStatusFail:    "fail",
+	ScoreStatusPass:    "pass",
+}
+
+func (ss ScoreStatus) String() string {
+	return statusName[ss]
+}
+
+// Score is the evaluation score that represents the result of an evaluator.
+// This struct includes information such as the score (numeric, string or other
+// types), the reasoning provided for this score (if any), the score status (if
+// any) and other details.
+type Score struct {
+	Id      string         `json:"id,omitempty"`
+	Score   any            `json:"score,omitempty"`
+	Status  string         `json:"status,omitempty" jsonschema:"enum=unknown,enum=fail,enum=pass"`
+	Error   string         `json:"error,omitempty"`
+	Details map[string]any `json:"details,omitempty"`
+}
+
+// EvaluationResult is the result of running the evaluator on a single Example.
+// An evaluator may provide multiple scores simultaneously (e.g. if they are using
+// an API to score on multiple criteria)
+type EvaluationResult struct {
+	TestCaseId string  `json:"testCaseId"`
+	TraceID    string  `json:"traceId,omitempty"`
+	SpanID     string  `json:"spanId,omitempty"`
+	Evaluation []Score `json:"evaluation"`
```

Comment:
Should this be a pointer? 

What is the general guideline for passing by reference / by value in golang?

Reply:
Hmm, I think I'll leave this as it is for now then. These don't make sense to be modified.